packages/hardware-ledger/src/transformers/certificates.ts

```diff
@@ -1,119 +1,122 @@
+            type: Ledger.CredentialParamsType.KEY_HASH
           };
       break;
     }
     case Cardano.CredentialType.ScriptHash:
     default: {
-      credential = {
-        scriptHashHex: certificate.stakeCredential.hash,
-        type: Ledger.StakeCredentialParamsType.SCRIPT_HASH
+      credentialParams = {
+        scriptHashHex: credential.hash,
+        type: Ledger.CredentialParamsType.SCRIPT_HASH
       };
     }
   }
 
-  return {
-    params: {
-      stakeCredential: credential
-    },
-    type
-  };
+  return credentialParams;
 };
 
-export const stakeDelegationCertificate: Transform<
-  Cardano.StakeDelegationCertificate,
-  Ledger.Certificate,
-  LedgerTxTransformerContext
-> = (certificate, context): Ledger.Certificate => {
-  const poolIdKeyHash = Cardano.PoolId.toKeyHash(certificate.poolId);
+const stakeCredentialMapper = (
+  credential: Cardano.Credential,
+  context: LedgerTxTransformerContext
+): Ledger.CredentialParams => {
   const knownAddress = context?.knownAddresses.find(
     (address) =>
-      Cardano.RewardAccount.toHash(address.rewardAccount) ===
-      (certificate.stakeCredential.hash as unknown as Crypto.Ed25519KeyHashHex)
+      Cardano.RewardAccount.toHash(address.rewardAccount) === (credential.hash as unknown as Crypto.Ed25519KeyHashHex)
   );
-  const rewardAddress = knownAddress ? Cardano.Address.fromBech32(knownAddress.rewardAccount)?.asReward() : null;
 
+  const rewardAddress = knownAddress ? Cardano.Address.fromBech32(knownAddress.rewardAccount)?.asReward() : null;
+  const path = util.stakeKeyPathFromGroupedAddress(knownAddress);
   const credentialType = rewardAddress
     ? rewardAddress.getPaymentCredential().type
-    : Ledger.StakeCredentialParamsType.SCRIPT_HASH;
+    : Ledger.CredentialParamsType.SCRIPT_HASH;
 
-  const path = util.stakeKeyPathFromGroupedAddress(knownAddress);
+  return credentialMapper(credential, credentialType, path);
+};
 
-  let credential: Ledger.StakeCredentialParams;
+const getStakeAddressCertificate = (
+  certificate: Cardano.StakeAddressCertificate,
+  context: LedgerTxTransformerContext,
+  type: Ledger.CertificateType.STAKE_REGISTRATION | Ledger.CertificateType.STAKE_DEREGISTRATION
+): Ledger.Certificate => ({
+  params: {
+    stakeCredential: stakeCredentialMapper(certificate.stakeCredential, context)
+  },
+  type
+});
 
-  switch (credentialType) {
-    case Cardano.CredentialType.KeyHash: {
-      credential = path
-        ? {
-            keyPath: path,
-            type: Ledger.StakeCredentialParamsType.KEY_PATH
-          }
-        : {
-            keyHashHex: certificate.stakeCredential.hash,
-            type: Ledger.StakeCredentialParamsType.KEY_HASH
-          };
-      break;
-    }
-    case Cardano.CredentialType.ScriptHash:
-    default: {
-      credential = {
-        scriptHashHex: certificate.stakeCredential.hash,
-        type: Ledger.StakeCredentialParamsType.SCRIPT_HASH
-      };
-    }
-  }
+const getNewStakeAddressCertificate = (
```

Comment:
```suggestion
const getNewStakeAddressCertificate: Transform<Cardano.NewStakeAddressCertificate, Ledger.Certificate, LedgerTxTransformerContext> = (
```

We implemented these Transform types to make sure all the mappings are done correctly (no field was skipped). So all mapping functions should use it

Reply:
@AngelCastilloB wouldn't it be just the same as specifying `Ledger.Certificate` as return type of the function?

Reply:
HI @vetalcore the transformer should be a bit more robust as it force all fields to be mapped, you can see how it works here https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/util/src/transformer.ts Previously we had bugs related to some objects partially mapped so @rhyslbw introduced this pattern, which should add some consistency to the mapping between the two types

Reply:
done here https://github.com/input-output-hk/cardano-js-sdk/pull/1211/commits/b866bb00a8812b0388a980e00ed35dd585fe696a